### PR TITLE
split platform-dock-layout

### PR DESCRIPTION
**Pull request by Sebastian-ubs**
This change is [Reviewable](https://reviewable.io/reviews/paranext/paranext-core/683)

### CODE REVIEW

File: src/renderer/components/docking/platform-dock-layout.component.tsx [hunk 1]
```diff
@@ -627,41 +66,42 @@ export default function PlatformDockLayout() {
...
+      const removedTab = dockLayoutRef.current.find(currentTabId) as RCDockTabInfo;
+      if ((removedTab.data as DialogData)?.isDialog && hasDialogRequest(currentTabId))
+        /* eslint-enable */
+        resolveDialogRequest(currentTabId, undefined, false);
+    }
+
+    (async () => {
+      if (onLayoutChangeRef.current) {
+        try {
+          await onLayoutChangeRef.current(...args);
```
Comment (Sebastian-ubs) [inline src/renderer/components/docking/platform-dock-layout.component.tsx:100]: Hi @tjcouch-sil could you have a look why this says
> A spread argument must either have a tuple type or be passed to a rest parameter.ts(2556)

File: src/renderer/components/docking/dock-layout-wrapper.component.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  onLayoutChange: (
+    newLayout: LayoutBase,
+    currentTabId?: string | undefined,
+    direction?: DropDirection | undefined,
+  ) => void;
+  defaultLayout?: LayoutData;
+  style?: CSSProperties;
+}>;
+
+const DockLayoutWrapper = forwardRef(function DockLayoutWrapper(
```
Comment (Sebastian-ubs) [inline src/renderer/components/docking/dock-layout-wrapper.component.tsx]: @tjcouch-sil: Is this correct? At least it renders :-)

**Review by Sebastian-ubs — COMMENTED** [review]
@tjcouch-sil could you please have a look at this?

**Comment by Sebastian-ubs** [comment]
If this approach looks good, I will also take care for the test code and thinking of moving / renaming the scss file.

**Review by tjcouch-sil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/683)* status: 0 of 8 files reviewed, 3 unresolved discussions (waiting on @Sebastian-ubs)

___
*[`src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx` line 20 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-Nlxd-PC7fPBFvdzicyZ:-Nlxd-PC7fPBFvdzicy_:b-2abfap) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx#L20)):*
> ```XML
> const DockLayoutWrapper = forwardRef(function DockLayoutWrapper(
>   { loadTab, saveTab, onLayoutChange, defaultLayout, style }: DockLayoutWrapperProps,
>   ref: LegacyRef<DockLayout> | undefined,
> ```

Though I am not familiar with it, I imagine `LegacyRef` likely shouldn't be used anymore. Have you tried `ForwardedRef`?

Here's the TypeScript React cheetsheet page on `forwardRef` for reference: https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/forward_and_create_ref/

**Review by tjcouch-sil — COMMENTED** [review]
Thanks for doing this! Yeah, the `forwardRef` idea you implemented here is great.

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/683#-:-NlyGlsO8aodvPvYZtIS:b-bkvoyi)* status: all files reviewed, 9 unresolved discussions (waiting on @Sebastian-ubs)

___
*[`src/renderer/components/docking/platform-dock-layout-sizing.tsx` line 1 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyE1Lm9inucFPT5XZ_:-NlyE1Lm9inucFPT5XZa:bs6mw5r) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-sizing.tsx#L1)):*
> ```XML
> import {
> ```

These new utility files need `.util.ts` at the end instead of `.tsx`: as mentioned in the [style guide "Filenames and paths" section](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-specific-guidelines), we need a `type` on the filename. Additionally, `.tsx` is used for React component files. These new files should hopefully be able to be changed into `.ts` files unless I'm missing something.
___
*[`src/renderer/components/docking/platform-dock-layout-sizing.tsx` line 10 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyEqW89sLwsSu66oAI:-NlyEqW98udt4VezvT0_:bxotdo8) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-sizing.tsx#L10)):*
> ```XML
> import { FloatPosition, FloatSize, LayoutSize, TabGroup } from 'rc-dock';
> import cloneDeep from 'lodash/cloneDeep';
> import DIALOGS from '../dialogs';
> ```

Here and in many places in this PR, please replace these relative paths with path aliases as mentioned in [the "Prefer path aliases for imports" line of the style guide](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-specific-guidelines).

For example, here, this import would instead be the following:
```typescript
import DIALOGS from '@renderer/components/dialogs';
```
___
*[`src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx` line 9 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyFcAe68Q7qgHJzbA1:-NlyFcAf9IqMnOh5WbRF:bluzj5d) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx#L9)):*
> ```XML
>   loadTab: (savedTabInfo: SavedTabInfo) => RCDockTabInfo;
>   saveTab: (dockTabInfo: RCDockTabInfo) => SavedTabInfo | undefined;
>   onLayoutChange: (
> ```

It may be best to use the `OnLayoutChangeRCDock` type here for consistency.
___
*[`src/renderer/components/docking/platform-dock-layout.component.tsx` line 80 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-Nly9JFG8wXFgeGQ7is9:-Nly9JFG8wXFgeGQ7isA:b-gn0lc2) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout.component.tsx#L80)):*
> ```XML
> }
> 
> function onLayoutChange(
> ```

This function's name is misleading; it does not indicate that it is returning a function but appears to be doing some execution. I recommend calling it `createOnLayoutChangeCallback` or something or preferably changing it back to being in-line since it is not used in multiple places, will have types implicitly, and will not require a name. Also, if you keep it split out, please move it above the function component according to the [style guide](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-module-declaration-order).
___
*[`src/renderer/components/docking/platform-dock-layout.component.tsx` line 85 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyC576-2cq7A5uNKEt:-NlyC576-2cq7A5uNKEu:b-pjl37b) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout.component.tsx#L85)):*
> ```XML
> ) {
>   /* eslint-disable-next-line */
>   return (...args: any[]) => {
> ```

You may be seeing the weird errors here because this is not typed. One disadvantage of moving this away from being in-line is that it does not come along with types. The correct types here would be `Parameters<OnLayoutChangeRCDock>`. And the return type on this function should be `ReturnType<OnLayoutChangeRCDock>`.

Please remove the `eslint-disable-next-line` as it is too generic. Hopefully we can solve the problem that led to adding that. If we can't solve the types issues, please use Intellisense to figure out the specific eslint rule to disable and explain the exception in a comment above it as specified in ["Explain rule exceptions" in the style guide](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-specific-guidelines)
___
*[`src/shared/models/docking-framework.model.ts` line 197 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyDVOA6w8If_WGAyM7:-NlyDVOA6w8If_WGAyM8:b-2qtsjf) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/shared/models/docking-framework.model.ts#L197)):*
> ```XML
> export type TabType = string;
> 
> export type RCDockTabInfo = TabData & TabInfo;
> ```

I would much prefer this type to stay internal to the docking framework files to avoid leaking out the `rc-dock` implementation details into more files than it already is in. Could we move these three additions (`TabType`, `RCDockTabInfo`, and `isTab`) back into the component file or into a single `platform-dock-layout.util.ts` file that encompasses all the new utility files to avoid having to export `RCDockTabInfo` and `TabType`?

**Review by Sebastian-ubs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/683)* status: 0 of 14 files reviewed, 9 unresolved discussions (waiting on @tjcouch-sil)

___
*[`src/renderer/components/docking/platform-dock-layout.component.tsx` line 80 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-Nly9JFG8wXFgeGQ7is9:-NmBRkqN3-44waQWd6TF:bb7f2mf) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout.component.tsx#L80)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

This function's name is misleading; it does not indicate that it is returning a function but appears to be doing some execution. I recommend calling it `createOnLayoutChangeCallback` or something or preferably changing it back to being in-line since it is not used in multiple places, will have types implicitly, and will not require a name. Also, if you keep it split out, please move it above the function component according to the [style guide](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-module-declaration-order).
</blockquote></details>

done
___
*[`src/renderer/components/docking/platform-dock-layout.component.tsx` line 85 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyC576-2cq7A5uNKEt:-NmBSTie6PE5CsgYBWNV:bb7f2mf) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout.component.tsx#L85)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

You may be seeing the weird errors here because this is not typed. One disadvantage of moving this away from being in-line is that it does not come along with types. The correct types here would be `Parameters<OnLayoutChangeRCDock>`. And the return type on this function should be `ReturnType<OnLayoutChangeRCDock>`.

Please remove the `eslint-disable-next-line` as it is too generic. Hopefully we can solve the problem that led to adding that. If we can't solve the types issues, please use Intellisense to figure out the specific eslint rule to disable and explain the exception in a comment above it as specified in ["Explain rule exceptions" in the style guide](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-specific-guidelines)
</blockquote></details>

done
___
*[`src/shared/models/docking-framework.model.ts` line 197 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyDVOA6w8If_WGAyM7:-NmBTHm5E8OkY4UM2oW8:bwzp2cr) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/shared/models/docking-framework.model.ts#L197)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I would much prefer this type to stay internal to the docking framework files to avoid leaking out the `rc-dock` implementation details into more files than it already is in. Could we move these three additions (`TabType`, `RCDockTabInfo`, and `isTab`) back into the component file or into a single `platform-dock-layout.util.ts` file that encompasses all the new utility files to avoid having to export `RCDockTabInfo` and `TabType`?
</blockquote></details>

done, moved into a spearate file *docking-framework-internal.model.ts*, but also reduced 3 others into *platform-dock-layout-storage.util.ts*
___
*[`src/renderer/components/docking/platform-dock-layout-sizing.tsx` line 1 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyE1Lm9inucFPT5XZ_:-NmBP77m1Ard4YaYGWYM:bkkqp54) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-sizing.tsx#L1)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

These new utility files need `.util.ts` at the end instead of `.tsx`: as mentioned in the [style guide "Filenames and paths" section](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-specific-guidelines), we need a `type` on the filename. Additionally, `.tsx` is used for React component files. These new files should hopefully be able to be changed into `.ts` files unless I'm missing something.
</blockquote></details>

done with splitting out *createRCDockTabFromTabInfo* into platform-dock-tab.component.tsx
___
*[`src/renderer/components/docking/platform-dock-layout-sizing.tsx` line 10 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyEqW89sLwsSu66oAI:-NmBPKnQ4oEyxq_0wt7I:bb7f2mf) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-sizing.tsx#L10)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Here and in many places in this PR, please replace these relative paths with path aliases as mentioned in [the "Prefer path aliases for imports" line of the style guide](https://github.com/paranext/paranext/wiki/Code-Style-Guide#typescript-specific-guidelines).

For example, here, this import would instead be the following:
```typescript
import DIALOGS from '@renderer/components/dialogs';
```
</blockquote></details>

done
___
*[`src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx` line 9 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-NlyFcAe68Q7qgHJzbA1:-NmBOmYq4QWA9LfOk1zF:bc1brxw) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx#L9)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

It may be best to use the `OnLayoutChangeRCDock` type here for consistency.
</blockquote></details>

This was not working with



_Code snippet:_
```XML
Type '(...args: any[]) => void' is not assignable to type 'OnLayoutChangeRCDock'.
  Type 'void' is not assignable to type 'Promise<void>'.ts(2322)
```
___
*[`src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx` line 20 at r1](https://reviewable.io/reviews/paranext/paranext-core/683#-Nlxd-PC7fPBFvdzicyZ:-NmBOyXU8FS3OGW6dZEK:bb7f2mf) ([raw file](https://github.com/paranext/paranext-core/blob/cc7e4e67566cae746ed0f45c73f2d63ee47205f1/src/renderer/components/docking/platform-dock-layout-wrapper.component.tsx#L20)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Though I am not familiar with it, I imagine `LegacyRef` likely shouldn't be used anymore. Have you tried `ForwardedRef`?

Here's the TypeScript React cheetsheet page on `forwardRef` for reference: https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/forward_and_create_ref/
</blockquote></details>

done

**Comment by Sebastian-ubs** [comment]
grafik

**Review by Sebastian-ubs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/683)* status: 15 of 26 files reviewed, 3 unresolved discussions (waiting on @tjcouch-sil)

___
*[`src/stories/mui-basics/snackbar.stories.tsx` line 2 at r4](https://reviewable.io/reviews/paranext/paranext-core/683#-NnFDnKs6HJH0X3eAnoc:-NnIwyN2CjCub4lNfVEo:b-4emfey) ([raw file](https://github.com/paranext/paranext-core/blob/35013ed961bfa6073e6a091be088c5f0e7ca28a4/src/stories/mui-basics/snackbar.stories.tsx#L2)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I see a number of stories got moved into a `mui-basics` folder. However, these files are not related to MUI (MUI happens to be an implementation detail at the current moment, but it is not necessarily going to be forever); these components are from `papi-components`. I recommend changing the name. Basics works fine. Or papi-components. Or just looping them in with platform. I don't know the relationship between this folder name and anything that displays on storybook, though, so please choose as you see fit :)
</blockquote></details>

That's to group them according to how they are named inside storybook (title). I've just changed this to "basics" now. This should include all basic components. We will for sure add more advanced components like dialogs or reusable layouts over time. We may event want to be able to add parts from extensions, but that's something to think about later where to host it so we can put as much components as possible together into this storybook as a workplace to develop and showcase components.

**Review by tjcouch-sil — APPROVED** [review]
:lgtm: thanks for the persistence and hard work on this!

> Reviewed 11 of 11 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/683#-:-NnJsoiE40bXxf-RF4Gc:b-q3hd3g)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

___
*[`src/stories/mui-basics/snackbar.stories.tsx` line 2 at r4](https://reviewable.io/reviews/paranext/paranext-core/683#-NnFDnKs6HJH0X3eAnoc:-NnJsi_v7yBP-feCwdtW:bgxid01) ([raw file](https://github.com/paranext/paranext-core/blob/35013ed961bfa6073e6a091be088c5f0e7ca28a4/src/stories/mui-basics/snackbar.stories.tsx#L2)):*
<details><summary><i>Previously, Sebastian-ubs wrote…</i></summary><blockquote>

That's to group them according to how they are named inside storybook (title). I've just changed this to "basics" now. This should include all basic components. We will for sure add more advanced components like dialogs or reusable layouts over time. We may event want to be able to add parts from extensions, but that's something to think about later where to host it so we can put as much components as possible together into this storybook as a workplace to develop and showcase components.
</blockquote></details>

Great :) thank you!

**Pull request merged**